File: backend/zev/iban.py
"""Small, dependency-free IBAN validation helpers."""

import re


def normalize_iban(value: str) -> str:
    """Return an IBAN in the canonical compact uppercase representation."""
    return re.sub(r"\s+", "", value).upper()


def is_valid_iban(value: str) -> bool:
    """Validate the IBAN structure and ISO 13616 MOD-97 checksum."""
    iban = normalize_iban(value)
    if not iban or not re.fullmatch(r"[A-Z]{2}[0-9]{2}[A-Z0-9]{11,30}", iban):
        return False
    rearranged = iban[4:] + iban[:4]
    numeric = "".join(
        str(ord(char) - ord("A") + 10) if char.isalpha() else char
        for char in rearranged
    )
    return int(numeric) % 97 == 1

File: frontend/src/lib/iban.ts
/** Return the compact uppercase representation used by the API. */
export function normalizeIban(value: string): string {
    return value.replace(/\s/g, '').toUpperCase()
}

export function isValidIban(value: string): boolean {
    const iban = normalizeIban(value)
    if (!iban || !/^[A-Z]{2}[0-9]{2}[A-Z0-9]{11,30}$/.test(iban)) return false
    const rearranged = `${iban.slice(4)}${iban.slice(0, 4)}`
    const numeric = [...rearranged].map((char) => /[A-Z]/.test(char)
        ? String(char.charCodeAt(0) - 55)
        : char).join('')
    let remainder = 0
    for (const chunk of numeric.match(/.{1,7}/g) ?? []) remainder = Number(`${remainder}${chunk}`) % 97
    return remainder === 1
}
